tests/frontend/test_engine.py

```diff
@@ -146,13 +157,211 @@ def inspect():
+        engine = StarshipEngine()
+        assert engine.client == mock_api_client()
+        assert engine.jobs == []
+
+    def test_reset(self, starship_engine):
+        """
+        Tests that StarshipEngine.jobs is correctly cleared when callling StarshipEngine.reset.
+        """
+        starship_engine.jobs.append(MagicMock())
+        assert len(starship_engine.jobs) == 1
+        starship_engine.reset()
+        assert len(starship_engine.jobs) == 0
+
+    def test__get_blackbird(self, starship_engine, monkeypatch):
+        """
+        Tests that StarshipEngine._get_blackbird returns the correct string given name,
+        shots, and program parameters.
+        """
+        methods = MagicMock()
+        inputs = MagicMock()
+
+        monkeypatch.setattr("strawberryfields.engine.to_blackbird", methods.to_blackbird)
+
+        output = starship_engine._get_blackbird(inputs.name, inputs.shots, inputs.program)
+
+        methods.to_blackbird.assert_called_once_with(inputs.program, version="1.0")
+        assert output._name == inputs.name
+        assert len(output._target.__setitem__.call_args_list) == 2
+        assert output._target.__setitem__.call_args_list[0] == call(
+            "name", starship_engine.backend_name
+        )
+        assert output._target.__setitem__.call_args_list[1] == call(
+            "options", {"shots": inputs.shots}
+        )
+
+    def test_queue_job(self, starship_engine, monkeypatch):
+        mock_job = MagicMock()
+        monkeypatch.setattr("strawberryfields.engine.Job", mock_job)
+        mock_job_content = MagicMock()
+
+        result = starship_engine._queue_job(mock_job_content)
+        mock_job.assert_called_once_with(client=starship_engine.client)
+        result.manager.create.assert_called_once_with(circuit=mock_job_content)
+        assert starship_engine.jobs == [result]
+
+    def test__run_program(self, starship_engine, monkeypatch):
+        """
+        Tests StarshipEngine._run_program. Asserts that a program is converted to blackbird code,
+        compiled into a job content string and that the job is queued. Also asserts that a
+        completed job's result samples are returned.
+        """
+        mock_to_blackbird = MagicMock()
+        mock__get_blackbird = MagicMock()
+        program = MagicMock()
+        mock_job = MagicMock()
+        mock_job.is_complete = True
+        mock_job.is_failed = False
+
+        monkeypatch.setattr("strawberryfields.engine.to_blackbird", mock_to_blackbird)
+        monkeypatch.setattr(starship_engine, "_get_blackbird", mock__get_blackbird)
+        monkeypatch.setattr(starship_engine, "_queue_job", lambda job_content: mock_job)
+
+        some_params = {"param": MagicMock()}
+        result = starship_engine._run_program(program, **some_params)
+
+        mock__get_blackbird.assert_called_once_with(program=program, param=some_params["param"])
+
+        assert result == mock_job.result.result.value
```

Comment:
how come `result` appears twice?

Reply:
`mock_job.result` is the resource (i.e. `job/<id>/result`). We have chosen to use `"result"` as a key in the response, so a `JobResult` resource (which can be accessed via `job.result` has a "result" attribute that is in fact the result. The response looks like `{"result": "..."}`.

Potentially a better key name could be "content".